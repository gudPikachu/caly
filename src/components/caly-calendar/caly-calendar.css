/**
 * @prop --grid: Specify grid template areas
 * @prop --font: Pass the font family you want the text to be in
 * @prop --border-width: Width of the calendar cell border
 * @prop --border-color: Color of the calendar cell border
 * @prop --cell-width: Width of the calendar cell
 * @prop --cell-height: Height of the calendar cell
 * @prop --selected-bg-color: Background color of selected day cell
 * @prop --other-month-visibility: Hidden by default, can be set to visible
 * @prop --other-month-border: None by default
 * @prop --hover-bg-color: Cell background color on hover
 * @prop --hover-color: Cell text color on hover
 */

:host {
  font-family: var(--font, -apple-system, BlinkMacSystemFont);
  user-select: none;
}

.grid {
  display: inline-grid;
  grid-template-areas: var(--grid,
    "nav nav"
    "mn mn"
  );
}

.navigation {
  grid-area: nav;
}

.inline-flex {
  display: inline-flex;
  flex-direction: column;
}

.flex {
  display: flex;
}

section {
  display: flex;
  justify-content: space-between;
}

.justify-center {
  justify-content: center;
}

.button {
  cursor: pointer;
}

table {
  border-collapse: collapse;
}

td {
  border: var(--border-width, 1px) solid var(--border-color, #e4e7e7);
  text-align: center;
  width: var(--cell-width, 36px);
  height: var(--cell-height, 36px);
}

td.borderless {
  border: none;
}

td.day-name {
  font-size: 12px;
}

td.today {
  color: #9e9c9c;
}

/* has to be after today to override it */
td.selected {
  background-color: var(--selected-bg-color, #00a699);
  color: white;
}

td.in-range {
  background-color: var(--in-range-bg-color, #00a699);
  color: white;
}

td.other-month {
  visibility: var(--other-month-visibility, hidden);
  border: var(--other-month-border, none); /* needed for Firefox */
}

td.day:not(.other-month):hover {
  background-color: var(--hover-bg-color, gainsboro);
  color: var(--hover-color, black);
  cursor: pointer;
}
