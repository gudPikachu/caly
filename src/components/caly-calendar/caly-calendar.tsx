import { Component, Prop, Event, EventEmitter, h, State } from '@stencil/core'
import {
  dayClass,
  selectedDayToCalendarDay,
  dayNames,
  monthName,
  range,
} from '../../utils/utils'
import { calendarMonth, IDay, getPreviousMonth, getNextMonth } from 'cntdys'

const chromeBordersFix = (table: HTMLElement) => {
  table.style.borderSpacing = table.style.borderSpacing === '0px' ? '' : '0px'
}

interface IMonth {
  year: number
  month: number
  weeks: IDay[][]
}

/**
 * @slot back – Slot for the previous month button
 * @slot forward – Slot for the next month button
 */
@Component({
  tag: 'caly-calendar',
  styleUrl: 'caly-calendar.css',
  shadow: true,
})
export class CalyCalendar {
  private tables: HTMLElement[] = []

  @State() hoverDay: IDay

  /** (required) Year (YYYY) */
  @Prop({ mutable: true, reflect: true }) year: number = new Date().getFullYear()
  /** (required) Month (1-12) */
  @Prop({ mutable: true, reflect: true }) month: number = new Date().getMonth() + 1
  /** (optional) Selected day (dd-mm-yyyy) */
  @Prop({ mutable: true, reflect: true }) selected: string
  /** (optional) Locale */
  @Prop() locale: string = 'en-US'
  /** (optional) Start of the week. 0 for Sunday, 1 for Monday, etc. */
  @Prop() startOfTheWeek: number = 0
  /** (optional) Number of months rendered */
  @Prop() numberOfMonths: number = 1
  /** (optional) Show previous number of months */
  @Prop() showPreviousNumberOfMonths: boolean = false
  /** (optional) Disabled days */
  @Prop() disableOnDay?: (timestamp: number) => boolean

  /** (optional) Range */
  @Prop() range: boolean = false
  /** (optional) Range start (dd-mm-yyyy) */
  @Prop({ mutable: true, reflect: true }) rangeStart: string
  /** (optional) Range end (dd-mm-yyyy) */
  @Prop({ mutable: true, reflect: true }) rangeEnd: string

  /** (optional) Event to listen for when new day is selected. */
  @Event({ eventName: 'daySelected' }) daySelected: EventEmitter
  /** (optional) Event to listen for when range start day is selected. */
  @Event({ eventName: 'rangeStartSelected' }) rangeStartSelected: EventEmitter
  /** (optional) Event to listen for when range end day is selected. */
  @Event({ eventName: 'rangeEndSelected' }) rangeEndSelected: EventEmitter

  private handleDayClick(day: IDay) {
    const dayInMonth = day.dayInMonth.toString().padStart(2, '0')
    const month = day.month.month.toString().padStart(2, '0')
    const selectedDay = `${day.month.year}-${month}-${dayInMonth}`

    if (this.range) {
      if (!this.rangeStart) {
        this.rangeStart = selectedDay
        this.rangeStartSelected.emit(selectedDay)
      } else if (!this.rangeEnd) {
        this.rangeEnd = selectedDay
        this.rangeEndSelected.emit(selectedDay)
      } else {
        this.rangeStart = selectedDay
        this.rangeEnd = null
        this.rangeStartSelected.emit(selectedDay)
      }
    } else {
      this.selected = selectedDay
      this.daySelected.emit(selectedDay)
    }
  }

  private handleMouseOver(day: IDay) {
    if (this.range) {
      this.hoverDay = day
    }
  }

  private back() {
    const { month, year } = getPreviousMonth(this.year, this.month)
    this.month = month
    this.year = year
    this.tables.forEach(table => chromeBordersFix(table))
  }

  private forward() {
    const { month, year } = getNextMonth(this.year, this.month)
    this.month = month
    this.year = year
    this.tables.forEach(table => chromeBordersFix(table))
  }

  render() {
    let month = { month: this.month, year: this.year }
    let months: IMonth[] = []

    for (let _i of range(this.numberOfMonths)) {
      let otherMonth = {
        year: month.year,
        month: month.month,
        weeks: calendarMonth(month.year, month.month, this.startOfTheWeek),
      }

      if (this.showPreviousNumberOfMonths) {
        months.unshift(otherMonth)
        month = getPreviousMonth(month.year, month.month) // going back
      } else {
        months.push(otherMonth)
        month = getNextMonth(month.year, month.month) // mutates month variable to progress it into next month
      }
    }

    return (
      <div class="grid">
        <section class="navigation">
          <div onClick={() => this.back()} class="button">
            <slot name="back">&lt;</slot>
          </div>
          <div onClick={() => this.forward()} class="button">
            <slot name="forward">&gt;</slot>
          </div>
        </section>

        {months.map(month => (
          <div class="inline-flex">
            <div class="flex justify-center">
              {monthName(month.year, month.month, this.locale)} {month.year}
            </div>
            <table
              ref={el => (this.tables.includes(el) ? {} : this.tables.push(el))}
            >
              <tr>
                {dayNames(this.startOfTheWeek, this.locale).map(dayName => (
                  <td class="borderless day-name">{dayName}</td>
                ))}
              </tr>
              {month.weeks.map(week => (
                <tr>
                  {week.map(day => (
                    <td
                      class={dayClass({
                        weekDay: day,
                        month: month.month,
                        year: month.year,
                        selectedDay: selectedDayToCalendarDay(this.selected),
                        rangeStart: selectedDayToCalendarDay(this.rangeStart),
                        rangeEnd: selectedDayToCalendarDay(this.rangeEnd),
                        hoverDay: this.hoverDay,
                        disableOnDay: this.disableOnDay,
                      })}
                      onClick={() => this.handleDayClick(day)}
                      onMouseOver={() => this.handleMouseOver(day)}
                    >
                      {day.dayInMonth}
                    </td>
                  ))}
                </tr>
              ))}
            </table>
          </div>
        ))}
      </div>
    )
  }
}
