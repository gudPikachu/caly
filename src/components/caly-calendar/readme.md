# caly-calendar



<!-- Auto Generated Below -->


## Usage

### Advanced

```html
<caly-calendar
  year="2020"
  month="1"
  selected="01-01-2020"
  locale="sl-SI"
  start-of-the-week="1"
>
  <!-- optional slots to change back and forward button style -->
  <div slot="back">
    <svg viewBox="0 0 1000 1000" style="width: 18px; height: 18px;">
      <path
        d="M336.2 274.5l-210.1 210h805.4c13 0 23 10 23 23s-10 23-23 23H126.1l210.1 210.1c11 11 11 21 0 32-5 5-10 7-16 7s-11-2-16-7l-249.1-249c-11-11-11-21 0-32l249.1-249.1c21-21.1 53 10.9 32 32z"
      />
    </svg>
  </div>
  <div slot="forward">
    <svg viewBox="0 0 1000 1000" style="width: 18px; height: 18px;">
      <path
        d="M694.4 242.4l249.1 249.1c11 11 11 21 0 32L694.4 772.7c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210.1-210.1H67.1c-13 0-23-10-23-23s10-23 23-23h805.4L662.4 274.5c-21-21.1 11-53.1 32-32.1z"
      />
    </svg>
  </div>
</caly-calendar>
```


### Simple

```html
<caly-calendar year="2020" month="1"></caly-calendar>
```



## Properties

| Property                     | Attribute                        | Description                                                    | Type                             | Default                     |
| ---------------------------- | -------------------------------- | -------------------------------------------------------------- | -------------------------------- | --------------------------- |
| `disableOnDay`               | --                               | (optional) Disabled days                                       | `(timestamp: number) => boolean` | `undefined`                 |
| `locale`                     | `locale`                         | (optional) Locale                                              | `string`                         | `'en-US'`                   |
| `month`                      | `month`                          | (required) Month (1-12)                                        | `number`                         | `new Date().getMonth() + 1` |
| `numberOfMonths`             | `number-of-months`               | (optional) Number of months rendered                           | `number`                         | `1`                         |
| `range`                      | `range`                          | (optional) Range                                               | `boolean`                        | `false`                     |
| `rangeEnd`                   | `range-end`                      | (optional) Range end (dd-mm-yyyy)                              | `string`                         | `undefined`                 |
| `rangeStart`                 | `range-start`                    | (optional) Range start (dd-mm-yyyy)                            | `string`                         | `undefined`                 |
| `selected`                   | `selected`                       | (optional) Selected day (dd-mm-yyyy)                           | `string`                         | `undefined`                 |
| `showPreviousNumberOfMonths` | `show-previous-number-of-months` | (optional) Show previous number of months                      | `boolean`                        | `false`                     |
| `startOfTheWeek`             | `start-of-the-week`              | (optional) Start of the week. 0 for Sunday, 1 for Monday, etc. | `number`                         | `0`                         |
| `year`                       | `year`                           | (required) Year (YYYY)                                         | `number`                         | `new Date().getFullYear()`  |


## Events

| Event                | Description                                                      | Type               |
| -------------------- | ---------------------------------------------------------------- | ------------------ |
| `daySelected`        | (optional) Event to listen for when new day is selected.         | `CustomEvent<any>` |
| `hoveredDay`         | (optional) Event to listen for what day is currently hovered.    | `CustomEvent<any>` |
| `rangeEndSelected`   | (optional) Event to listen for when range end day is selected.   | `CustomEvent<any>` |
| `rangeStartSelected` | (optional) Event to listen for when range start day is selected. | `CustomEvent<any>` |


## Slots

| Slot        | Description                                          |
| ----------- | ---------------------------------------------------- |
| `"back"`    | Slot for the previous month button                   |
| `"forward"` | Slot for the next month button                       |
| `"misc"`    | Slot for the miscellaneous, e.g. date preset buttons |


## CSS Custom Properties

| Name                       | Description                                         |
| -------------------------- | --------------------------------------------------- |
| `--cell-border-color`      | Color of the calendar cell border                   |
| `--cell-border-style`      | Style of the calendar cell border                   |
| `--cell-border-width`      | Width of the calendar cell border                   |
| `--cell-font-size`         | Font size of the calendar cell                      |
| `--cell-height`            | Height of the calendar cell                         |
| `--cell-text-align`        | Text alignment of the calendar cell                 |
| `--cell-width`             | Width of the calendar cell                          |
| `--day-cursor`             | Cursor on cell hover                                |
| `--day-name-font-size`     | Font size of day name                               |
| `--font`                   | Pass the font family you want the text to be in     |
| `--grid`                   | Specify grid template areas                         |
| `--grid-column-gap`        | Specify grid column gap                             |
| `--hover-bg-color`         | Cell background color on hover                      |
| `--hover-color`            | Cell text color on hover                            |
| `--in-range-bg-color`      | Background color of in-range cell                   |
| `--in-range-color`         | Color of in-range cell                              |
| `--navigation-height`      | Specify grid column gap                             |
| `--other-month-border`     | None by default                                     |
| `--other-month-visibility` | Hidden by default, can be set to visible            |
| `--out-of-range-color`     | Cell text color when day is out of range / disabled |
| `--selected-bg-color`      | Background color of selected day cell               |
| `--selected-color`         | Color of selected day cell                          |
| `--today-color`            | Font color of today's cell                          |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
