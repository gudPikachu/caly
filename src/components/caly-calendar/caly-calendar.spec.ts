import { newSpecPage } from '@stencil/core/testing'
import { CalyCalendar } from './caly-calendar'

it('should render caly-calendar', async () => {
  const page = await newSpecPage({
    components: [CalyCalendar],
    html: `<caly-calendar year="2010" month="1" selected="2010-01-01"></caly-calendar>`,
  })

  expect(page.root).toMatchSnapshot('render-2010-01-01')
})
