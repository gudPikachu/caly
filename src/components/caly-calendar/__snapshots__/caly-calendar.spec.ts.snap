// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render caly-calendar: render-01-01-2010 1`] = `
<caly-calendar month="1" selected="01-01-2010" year="2010">
  <mock:shadow-root>
    <div class="grid">
      <section class="navigation">
        <div class="button">
          <slot name="back">
            &lt;
          </slot>
        </div>
        <div class="button">
          <slot name="forward">
            &gt;
          </slot>
        </div>
      </section>
      <div class="inline-flex">
        <div class="flex justify-center">
          January 2010
        </div>
        <table>
          <tr>
            <td class="borderless day-name">
              Su
            </td>
            <td class="borderless day-name">
              Mo
            </td>
            <td class="borderless day-name">
              Tu
            </td>
            <td class="borderless day-name">
              We
            </td>
            <td class="borderless day-name">
              Th
            </td>
            <td class="borderless day-name">
              Fr
            </td>
            <td class="borderless day-name">
              Sa
            </td>
          </tr>
          <tr>
            <td class="day other-month weekend">
              27
            </td>
            <td class="day other-month">
              28
            </td>
            <td class="day other-month">
              29
            </td>
            <td class="day other-month">
              30
            </td>
            <td class="day other-month">
              31
            </td>
            <td class="current-month day selected">
              1
            </td>
            <td class="current-month day weekend">
              2
            </td>
          </tr>
          <tr>
            <td class="current-month day weekend">
              3
            </td>
            <td class="current-month day">
              4
            </td>
            <td class="current-month day">
              5
            </td>
            <td class="current-month day">
              6
            </td>
            <td class="current-month day">
              7
            </td>
            <td class="current-month day">
              8
            </td>
            <td class="current-month day weekend">
              9
            </td>
          </tr>
          <tr>
            <td class="current-month day weekend">
              10
            </td>
            <td class="current-month day">
              11
            </td>
            <td class="current-month day">
              12
            </td>
            <td class="current-month day">
              13
            </td>
            <td class="current-month day">
              14
            </td>
            <td class="current-month day">
              15
            </td>
            <td class="current-month day weekend">
              16
            </td>
          </tr>
          <tr>
            <td class="current-month day weekend">
              17
            </td>
            <td class="current-month day">
              18
            </td>
            <td class="current-month day">
              19
            </td>
            <td class="current-month day">
              20
            </td>
            <td class="current-month day">
              21
            </td>
            <td class="current-month day">
              22
            </td>
            <td class="current-month day weekend">
              23
            </td>
          </tr>
          <tr>
            <td class="current-month day weekend">
              24
            </td>
            <td class="current-month day">
              25
            </td>
            <td class="current-month day">
              26
            </td>
            <td class="current-month day">
              27
            </td>
            <td class="current-month day">
              28
            </td>
            <td class="current-month day">
              29
            </td>
            <td class="current-month day weekend">
              30
            </td>
          </tr>
          <tr>
            <td class="current-month day weekend">
              31
            </td>
            <td class="day other-month">
              1
            </td>
            <td class="day other-month">
              2
            </td>
            <td class="day other-month">
              3
            </td>
            <td class="day other-month">
              4
            </td>
            <td class="day other-month">
              5
            </td>
            <td class="day other-month weekend">
              6
            </td>
          </tr>
        </table>
      </div>
    </div>
  </mock:shadow-root>
</caly-calendar>
`;
